### PR TITLE
support Range header for rest block api

### PR DESCRIPTION
**Pull request by Ljzn**
Hi, folks. This is an implementation of HTTP Range header for the `/rest/block/...bin` api. As the block size got bigger and bigger, we met some problems in downloading the entire block data from the node. The connection keeps dropping when downloading a 1G~3G block. While we trying to improve our network, a HTTP Range header support would be a nice fix for this.

For example, a `Range: 0-1024000` header can be set in the request to download the first 1M data of block. As well any other part. It effectively reduces the failure rate.

The code works fine in our production. I'm not a C++ professional, so it's just a suggestion on how to implement this feature.

### CODE REVIEW

**Comment by pfromberg** [comment]
Thanks for your contribution. We are evaluating it now.

**Comment by CassClark** [comment]
Internal ticker: SVN-2006

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -1277,26 +1277,75 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
...
+    uint64_t offset;
+    uint64_t content_length;
+    std::unique_ptr<CForwardReadonlyStream> stream;
+
+    if (!hasDiskBlockMetaData && hasRangeHeader) {
+        stream = blockIndex.StreamSyncBlockFromDisk();
+        do
+        {
+            auto chunk = stream->Read(4096);
+            metadata.diskDataSize += chunk.Size();
```
Comment (richmills3) [inline src/rpc/blockchain.cpp:1295]: Why do we need to update the `CDiskBlockMetaData` here? Idn't the meta-data correct at the time we fetch it (above)?

File: test/functional/rest.py [hunk 2]
```diff
@@ -252,6 +261,13 @@ def run_test(self):
         assert_greater_than(int(response.getheader('content-length')), 80)
         response_str = response.read()
 
+        # get binary block with range
+        response = http_get_call_with_headers(
+            url.hostname, url.port, '/rest/block/' + bb_hash + self.FORMAT_SEPARATOR + "bin", {'Range': '10-29'}, True)
```
Comment (richmills3) [inline test/functional/rest.py:266]: Is this compliant with the standard? I'm not certain, but by my reading of the RFC you must also include the units of the range being requested, which in this case would be `bytes=`.

https://www.rfc-editor.org/rfc/rfc9110.html#name-range-requests

Comment (Ljzn) [reply test/functional/rest.py:266]: @richmills3 You're right. It should start with a unit like `bytes=`, I forgot when I implemented this.
MDN has an example too https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range

Comment (richmills3) [reply test/functional/rest.py:266]: Further reading of the spec (https://httpwg.org/specs/rfc9110.html#field.content-range) makes me also think the response code should be 206 in this case, do you agree?

I think for us to take this change I'd like to see:

1.  `bytes=` unit added to the request parsing.
2. Response code 206 for range responses (unless I'm wrong about that).
3. More tests, in particular some negative tests, eg; for range requests outside of block, bad request format.
4. Also see my comment on `block_index.cpp`.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -1277,26 +1277,75 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
...
+        } while (!stream->EndOfStream());
+
+        blockIndex.SetBlockIndexFileMetaDataIfNotSet(metadata, mapBlockIndex);
```
Comment (richmills3) [inline src/rpc/blockchain.cpp:1298]: Do we need to do this metadata calculation? If I'm understanding correctly we only need the metadata to report the `Content-Length` and `Contents-Range` back. If we don't have the metadata then couldn't we just skip the `Content-Length` field (as the old code does) and report the `Content-Range` using `/*` (as it says here: https://httpwg.org/specs/rfc9110.html#field.content-range)?

File: test/functional/rest.py (hunk 2 above)
Comment (Ljzn) [reply test/functional/rest.py:266]: I totally agree, these changes are needed and great. 

I'm trying to deal with some issues with building the project. There are some strange errors during the `make`(https://github.com/Ljzn/bitcoin-sv/actions/runs/3800283694/jobs/6463746175).  Maybe I should rebase my branch to the latest code on the master. Anyway, I will start coding on these as long as I find some free time and solve the build problem.

Comment (richmills3) [reply test/functional/rest.py:266]: I think you're hitting some known build issues on later OSs (Ubuntu 22?). We have fixes for those that will be in the next release but I don't think they're in master yet. They are easy to fix though, just `#include <mutex>` as the compiler suggests.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -1277,26 +1277,70 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
...
+                    }
+                    contentLen = re - rs + 1;
+
+                    if (hasDiskBlockMetaData) {
+                        if (rs >= metadata.diskDataSize)
+                        {
+                            throw block_parse_error("Invalid Range parameter, start >= data_size");
+                        }
+
+                        offset = rs;
```
Comment (richmills3) [inline src/rpc/blockchain.cpp:1325]: If we don't have the block meta-data then `offset` will be undefined when we come to use it `req.WriteHeader` below. Should it perhaps be set above at the same time as `contentLen`?

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -1277,26 +1277,70 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
...
+                {
+                    std::string s = range_header.second;
+                    if (s.find("bytes=") != 0)
+                    {
+                        throw block_parse_error("Invalid Range header format, should starts with 'bytes='");
+                    }
+                    s.erase(0, 6);
+                    std::string delimiter = "-";
+                    std::string rs_s = s.substr(0, s.find(delimiter));
+                    s.erase(0, s.find(delimiter) + delimiter.length());
```
Comment (richmills3) [inline src/rpc/blockchain.cpp:1307]: What happens here (and just above) if `delimiter` is not found (range string is badly formatted)? I think it should explicitly check for that.

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1277,26 +1277,70 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
...
+                        uint64_t remain = metadata.diskDataSize - offset;
+                        contentLen = std::min(remain, contentLen);
+                        totalLen = std::to_string(metadata.diskDataSize);
+                    }
+
+                    req.WriteHeader("Content-Length", std::to_string(contentLen));
+                    req.WriteHeader("Content-Range", "bytes " + std::to_string(offset) + "-" + 
+                        std::to_string(contentLen - 1) + "/" + totalLen);
+                }
+                catch (...) {
```
Comment (richmills3) [inline src/rpc/blockchain.cpp:1336]: We will catch the `block_parse_error`s thrown above here as well, and then we will loose the extra information in their exception strings when we throw again just below. Maybe it would be better to also add here:
````
catch(const block_parse_error&)
{
    throw;
}
````

**Comment by richmills3** [comment]
FYI; I'm fixing my final code review comments myself here locally for you and will merge this change to our local release branch. It will be included in the next public release.

**Comment by Ljzn** [comment]
> FYI; I'm fixing my final code review comments myself here locally for you and will merge this change to our local release branch. It will be included in the next public release.

Excited to hear that! I'm on holiday. Appreciate all your work and reviews. It's a wonderful journey to contribute to bitcoin-sv.
